wled00/espnow_broadcast.cpp

```diff
@@ -0,0 +1,329 @@
+
+#ifndef WLED_DISABLE_ESPNOW_NEW
+#include <espnow_broadcast.h>
+
+#if defined ESP32
+#include <WiFi.h>
+
+#elif defined ESP8266
+#include <ESP8266WiFi.h>
+#define WIFI_MODE_STA WIFI_STA
+#else
+#error "Unsupported platform"
+#endif //ESP32
+
+#include <esp_wifi.h>
+#include <esp_wifi_types.h>
+#include <esp_now.h>
+#include <esp_idf_version.h>
+#include <freertos/ringbuf.h>
+
+#if ESP_IDF_VERSION < ESP_IDF_VERSION_VAL(4, 0, 0)
+// Legacy Event Loop
+ESP_EVENT_DEFINE_BASE(SYSTEM_EVENT);
+#define WIFI_EVENT SYSTEM_EVENT
+#define WIFI_EVENT_STA_START SYSTEM_EVENT_STA_START
+#define WIFI_EVENT_STA_STOP SYSTEM_EVENT_STA_STOP
+#define WIFI_EVENT_AP_START SYSTEM_EVENT_AP_START
+#endif
+
+//#define ESPNOW_DEBUGGING
+
+#define BROADCAST_ADDR_ARRAY_INITIALIZER {0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}
+#define WLED_ESPNOW_WIFI_CHANNEL 1
+
+typedef struct {
+    uint8_t mac[6];
+    uint8_t len;
+    uint8_t data[WLED_ESPNOW_MAX_MESSAGE_LENGTH];
+} QueuedNetworkMessage;
+static_assert(WLED_ESPNOW_MAX_MESSAGE_LENGTH <= ESP_NOW_MAX_DATA_LEN, "WLED_ESPNOW_MAX_MESSAGE_LENGTH must be <= 250 bytes");
+
+class QueuedNetworkRingBuffer {
+  protected:
+    //QueuedNetworkMessage messages[WLED_ESPNOW_MAX_QUEUED_MESSAGES];
+    RingbufHandle_t buf = nullptr;
+
+  public:
+    QueuedNetworkRingBuffer() {
+        buf = xRingbufferCreateNoSplit(sizeof(QueuedNetworkMessage), WLED_ESPNOW_MAX_QUEUED_MESSAGES);
+    }
+
+    bool push(const uint8_t* mac, const uint8_t* data, uint8_t len) {
+#if ESP_IDF_VERSION < ESP_IDF_VERSION_VAL(4, 0, 0)
+        QueuedNetworkMessage msg[1];
```

Comment:
this could be optimized in the future by writing our own ringbuffer for the old IDF if desired